find_exponential/src/main.rs

```diff
@@ -0,0 +1,197 @@
+use std::fs;
+use std::path::Path;
+use std::collections::HashSet;
+use std::time::{Duration, Instant};
+use std::panic;
+use std::sync::mpsc;
+use std::env;
+use std::io::{self, BufRead};
+
+use walkdir::WalkDir;
+use itertools::Itertools;
+use pulldown_cmark::{Parser, Options};
+use ndarray::Array2;
+use rand::seq::SliceRandom;
+use threadpool::ThreadPool;
+
+mod literals;
+mod black_box;
+mod scoring;
+
+/// How many combinations are batch-processed.
+const BATCH_SIZE: usize = 10_000;
+/// Combination length.
+const COMBINATIONS: usize = 4;
+/// If parsing takes longer than this many milliseconds, it is assumed to be non-linear.
+const MAX_MILLIS: u128 = 500;
+/// Size of repetition in bytes to parse.
+const NUM_BYTES: usize = 256*1024;
+/// Number of samples per pattern tested.
+const SAMPLE_SIZE: usize = 5;
+/// Score function to use when figuring out if something is non-linear.
+/// `scoring::{slope_stddev,pearson_correlation}`
+const SCORE_FUNCTION: fn(&Array2<f64>) -> (f64, bool) = scoring::slope_stddev;
+/// If slope_stddev is used, if the standard deviation is larger than this, it's assumed to be non-linear.
+const ACCEPTANCE_STDDEV: f64 = 20.0;
+/// If pearson_correlation is used, if the correlation coefficient is below this value,
+/// it's assumed to be non-linear.
+const ACCEPTANCE_CORRELATION: f64 = 0.995;
+/// 0 / 1 / 2 / 3
+const DEBUG_LEVEL: u8 = 2;
+
+fn main() {
+    let num_cpus = num_cpus::get() * 3 / 4;
+
+    let arg = env::args().nth(1);
+    if let Some("--retest") = arg.as_ref().map(|s| s.as_str()) {
+        for pattern in io::stdin().lock().lines() {
+            let pattern = pattern.unwrap();
+            println!("retesting {:?}", pattern);
+            test(&pattern);
+        }
+        return;
+    } else {
+        fuzz(num_cpus);
+    }
+}
+
+fn fuzz(num_cpus: usize) {
+    // get all literals from pulldown-cmark's source code
+    let walkdir = WalkDir::new("../src")
```

Comment:
Currently I just extract all literals. I definitely agree that that's probably not the best, especially since there are a lot of printed / debug string literals. I don't think that a fully manual list is the correct answer though. While it might seem nice at first, changes to pulldown-cmark won't be reflected in the token list automatically and we'll need to manually keep it updated.

I may try just extracting literals from if-conditions and match arms. I'll compare those to a manually trimmed down list to see if all relevant tokens are included. Additionally, we probably want to add some literals manually after the automatic scan.

Reply:
These comments are superseded by #302, which was fixed in 747a765.